### ISSUE TITLE
4+ log-forward sections causes segmentation error on startup

### ISSUE THREAD

**Issue by JanPMadsen** [haproxy/haproxy#1464]
### Detailed Description of the Problem

Hi
When trying to add more that 3 log-forward sections to be anble to handle requests on multiple UDP ports and take different actions on them, HAProxy causes segmentation error:

I'm using this configuration section:

```
log-forward syslogUDPFrontend-514
        # Listen on UDP and TCP 514
        dgram-bind "${vipAddr}":514
        log ring@syslogRing-514 local0

log-forward syslogUDPFrontend-10514
        # Listen on UDP and TCP 10514 
        dgram-bind "${vipAddr}":10514
        log ring@syslogRing-10514 local1

log-forward syslogUDPFrontend-10515
        # Listen on UDP and TCP 10515 
        dgram-bind "${vipAddr}":10515
        log ring@syslogRing-10515 local2

log-forward syslogUDPFrontend-10516
        # Listen on UDP and TCP 10516 
        dgram-bind "${vipAddr}":10516
        log ring@syslogRing-10516 local0
```

A configuration check does not complain.
But system crashes with:

### Expected Behavior

I just wanted to be able to create all the log-forward sections I want :) or more than 3 which seems to be the limit at current time.
At least 20

### Steps to Reproduce the Behavior

Edit /etc/haproxy.cfg

Add the following configuration lines:

```
#---------------------------------------------------------------------
# UDP Listeners for sylog messages
#---------------------------------------------------------------------

log-forward UDP514
        # Listen on UDP and TCP 514
        dgram-bind "${vipAddr}":514
        log ring@syslogRing-514 local0 info

log-forward UDP10514
        # Listen on UDP and TCP 10514 
        dgram-bind "${vipAddr}":10514
        log ring@syslogRing-10514 local0 info

log-forward UDP10515
        # Listen on UDP and TCP 10515 
        dgram-bind "${vipAddr}":10515
        log ring@syslogRing-10515 local0 info

log-forward UDP10516
        # Listen on UDP and TCP 10516 
        dgram-bind "${vipAddr}":10516
        log ring@syslogRing-10516 local0 info

log-forward UDP10517
        # Listen on UDP and TCP 10517
        dgram-bind "${vipAddr}":10517
        log ring@syslogRing-10517 local0 info
#########################
# Rings to handle UDP connections, and to upscale UDP to TCP to be able to make loadbalancing on received UDP data and backend server check.
############
ring syslogRing-514
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog-514 "${vipAddr}":514 check log-proto octet-count

ring syslogRing-10514
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog-10514 "${vipAddr}":10514 check log-proto octet-count

ring syslogRing-10515 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        maxlen 2400
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog-10515 "${vipAddr}":10515 check log-proto octet-count

ring syslogRing-10516
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        maxlen 2400
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog-10516 "${vipAddr}":10516 check log-proto octet-count

ring syslogRing-10517
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        maxlen 2400
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog-10517 "${vipAddr}":10517 check log-proto octet-count
#################
# Frontend TCP Listeners
#################
frontend syslogTCPFrontend-514
        bind "${vipAddr}":514
        default_backend syslogTCPBackend-514

frontend syslogTCPFrontend-10514
        bind "${vipAddr}":10514
        default_backend syslogTCPBackend-10514

frontend syslogTCPFrontend-10515
        bind "${vipAddr}":10515
        default_backend syslogTCPBackend-10515

frontend syslogTCPFrontend-10516
        bind "${vipAddr}":10516
        default_backend syslogTCPBackend-10516

frontend syslogTCPFrontend-10517
        bind "${vipAddr}":10517
        default_backend syslogTCPBackend-10517

frontend syslogTCPFrontend-10518
        bind "${vipAddr}":10518
        default_backend syslogTCPBackend-10518

#######################
# Backend servers
#######################

backend syslogTCPBackend-514
        balance roundrobin
        default-server check
        server syslogB1-514 "${syslog1}":514
        server syslogB2-514 "${syslog2}":514

backend syslogTCPBackend-10514
        balance roundrobin
        default-server check
        server syslogB1-10514 "${syslog1}":10514
        server syslogB2-10514 "${syslog2}":10514

backend syslogTCPBackend-10515
        balance roundrobin
        default-server check
        server syslogB1-10515 "${syslog1}":10515
        server syslogB2-10515 "${syslog2}":10515

backend syslogTCPBackend-10516
        balance roundrobin
        default-server check
        server syslogB1-10516 "${syslog1}":10516
        server syslogB2-10516 "${syslog2}":10516

backend syslogTCPBackend-10517
        balance roundrobin
        default-server check
        server syslogB1-10517 "${syslog1}":10517
        server syslogB2-10517 "${syslog2}":10517

backend syslogTCPBackend-10518
        balance roundrobin
        default-server check
        server syslogB1-10518 "${syslog1}":10518
        server syslogB2-10518 "${syslog2}":10518
```

### Do you have any idea what may have caused this?

NO !

### Do you have an idea how to solve the issue?

No really !

### What is your configuration?

```haproxy
log         127.0.0.1 local2

    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     8000
    user        haproxy
    group       haproxy
    daemon

    # turn on stats unix socket
    stats socket /var/lib/haproxy/stats

    # Configure environment variables. This must be change accordingly to the setup of these log servers. 
    setenv      vipAddr "10.10.10.1"
    setenv      syslog1 "10.10.10.2"
    setenv      syslog2 "10.10.10.3"



#---------------------------------------------------------------------
# common defaults that all the 'listen' and 'backend' sections will
# use if not designated in their block
#---------------------------------------------------------------------
defaults
    mode                    tcp
    log                     global
    option                  tcplog
    option                  dontlognull
    option                  redispatch
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          5s
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 10000


#---------------------------------------------------------------------
# UDP Listeners for sylog messages
#---------------------------------------------------------------------
#
log-forward syslogUDPFrontend-514
        # Listen on UDP and TCP 514
        dgram-bind "${vipAddr}":514
        log ring@syslogRing-514 local0 info

log-forward syslogFrontend-10514
        # Listen on UDP and TCP 10514 
        dgram-bind "${vipAddr}":10514
        log ring@syslogRing-10514 local0 info

log-forward syslogFrontend-10515
        # Listen on UDP and TCP 10515 
        dgram-bind "${vipAddr}":10515
        log ring@syslogRing-10515 local0 info

log-forward syslogFrontend-10517
        # Listen on UDP and TCP 10517 
        dgram-bind "${vipAddr}":10517
        log ring@syslogRing-10517 local0 info

log-forward syslogFrontend-10518
        # Listen on UDP and TCP 10518 
        dgram-bind "${vipAddr}":10518
        log ring@syslogRing-10518 local0 info

log-forward syslogFrontend-10519
        # Listen on UDP and TCP 10519 
        dgram-bind "${vipAddr}":10519
        log ring@syslogRing-10519 local0 info

log-forward syslogFrontend-10520
        # Listen on UDP and TCP 10520
        dgram-bind "${vipAddr}":10520
        log ring@syslogRing-10520 local0 info

#########################
# Rings to handle UDP connections, and to upscale UDP to TCP to be able to make loadbalancing on received UDP data and backend server check.
############
ring syslogRing-514 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":514 check log-proto octet-count

ring syslogRing-10514 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10514 check log-proto octet-count

ring syslogRing-10515 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        #format rfc5424
        maxlen 2400
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10515 check log-proto octet-count

ring syslogRing-10517 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10517 check log-proto octet-count

ring syslogRing-10518 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10518 check log-proto octet-count

ring syslogRing-10519 
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10519 check log-proto octet-count

ring syslogRing-10520
        description "My local log buffer, all messages are sent to all servers here, and can keep some in buffer if backed server is down"
        format rfc5424
        maxlen 1200
        size 32764000
        timeout connect 5s
        timeout server 10s
        server syslog1 "${vipAddr}":10520 check log-proto octet-count

#################
# Frontend TCP Listeners
#################
frontend syslogTCPFrontend-514
        bind "${vipAddr}":514
        default_backend syslogTCPBackend-514

frontend syslogTCPFrontend-10514
        bind "${vipAddr}":10514
        default_backend syslogTCPBackend-10514

frontend syslogTCPFrontend-10515
        bind "${vipAddr}":10515
        default_backend syslogTCPBackend-10515

frontend syslogTCPFrontend-10517
        bind "${vipAddr}":10517
        default_backend syslogTCPBackend-10517

frontend syslogTCPFrontend-10518
        bind "${vipAddr}":10518
        default_backend syslogTCPBackend-10518

frontend syslogTCPFrontend-10519
        bind "${vipAddr}":10519
        default_backend syslogTCPBackend-10519

frontend syslogTCPFrontend-10520
        bind "${vipAddr}":10520
        default_backend syslogTCPBackend-10520


#######################
# Backend servers
#######################

backend syslogTCPBackend-514
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":514 
        server syslog2 "${syslog2}":514 

backend syslogTCPBackend-10514
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10514 
        server syslog2 "${syslog2}":10514 

backend syslogTCPBackend-10515
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10515 
        server syslog2 "${syslog2}":10515 

backend syslogTCPBackend-10517
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10517 
        server syslog2 "${syslog2}":10517 

backend syslogTCPBackend-10518
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10518
        server syslog2 "${syslog2}":10518 

backend syslogTCPBackend-10519
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10519
        server syslog2 "${syslog2}":10519 

backend syslogTCPBackend-10520
        balance roundrobin
        default-server check 
        server syslog1 "${syslog1}":10520
        server syslog2 "${syslog2}":10520
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.4-acb1d0b 2021/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.4.html
Running on: Linux 4.18.0-193.28.1.el8_2.x86_64 #1 SMP Fri Oct 16 13:38:49 EDT 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=20).
Built with OpenSSL version : OpenSSL 1.1.1c FIPS  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.1c FIPS  28 May 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.1 20191121 (Red Hat 8.3.1-5)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
# haproxy -f /etc/haproxy/haproxy.cfg -db -d 
Note: setting global.maxconn to 131020.
Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result FAILED
Total: 3 (2 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
Using epoll() as the polling mechanism.
Segmentation fault
```

### Additional Information

This is all running on RHEL 8

Currenlty only 1 firewall is sending syslog to the system

**Comment by EmericBr** [comment]
I have a reproducer here with your config (it segfaults receiving the first udp log). I'm debugging

**Comment by EmericBr** [comment]
[0001-BUG-MAJOR-segfault-using-multiple-log-forward-sectio.patch.txt](https://github.com/haproxy/haproxy/files/7633428/0001-BUG-MAJOR-segfault-using-multiple-log-forward-sectio.patch.txt)
This patch should fix the issue.

**Comment by JanPMadsen** [comment]
Nice one 😊

May I ask how to deploy this patch ?
I see:
diff --git a/src/log.c b/src/log.c
index ecf771221..92eef6ebe 100644
--- a/src/log.c
+++ b/src/log.c
@@ -3738,10 +3738,9 @@ int cfg_parse_log_forward(const char *file, int linenum, char **args, int kwm)
                                                                                      goto out;
                                                         }
+                                                       init_new_proxy(px);
                                                         px->next = cfg_log_forward;
                                                         cfg_log_forward = px;
-
-                                                        init_new_proxy(px);
                                                         px->conf.file = strdup(file);
                                                         px->conf.line = linenum;
                                                         px->mode = PR_MODE_SYSLOG;

But what should I do to fix my currently setup (2.4.4)
New build / installation ?

Best regards
/Jan

**Comment by EmericBr** [comment]
if you are not able to recompile notice It will be merged in master asap, and it will be backported until v2.3 with other fixes

**Comment by EmericBr** [comment]
Please also notice that your configuration won't perform a per message load balancing because as soon a connection is establised between a ring and the frontend, it will remain maintained and all the messages will be pushed to the same server through the frontend/backend tcp proxy since the syslog tcp session is persistent

**Comment by EmericBr** [comment]
If you want to perform per message lb take a look at documentation example:
    log-forward sylog-loadb
        dgram-bind 127.0.0.1:1514
        bind 127.0.0.1:1514
        # all messages on stderr
        log global
        # all messages on local tcp syslog server
        log ring@myring local0
        # load balance messages on 4 udp syslog servers
        log 127.0.0.1:10001 sample 1:4 local0
        log 127.0.0.1:10002 sample 2:4 local0
        log 127.0.0.1:10003 sample 3:4 local0
        log 127.0.0.1:10004 sample 4:4 local0

**Comment by capflam** [comment]
now pushed, thanks @EmericBr !

**Comment by JanPMadsen** [comment]
Hi.
Hmm
Okay. We did the ring part to be able to check if the backend syslog servers was alive. (They listen both on UDP and TCP)
Since HAProxy does not support any UDP checks that was needed !

A simple script execution running to check backend UDP server:
#!/bin/bash
/usr/bin/nc -uz -w1 $1 $2
exit $?

Best regards
Jan

**Comment by wtarreau** [comment]
Be careful with UDP checks. **They're not reliable at all**. What you're doing above consists in sending two 1-byte packets 1 second apart to the log server. First it may cause your log server to emit an empty log line for each of them. But more importantly, netcat will rely on the receipt of an ICMP port unreachable packet in response to decide that the port is closed. This means that while there are little risks to see a working server being reported down, there are high risks of seeing a down server reported up. For example if the target machine is dead you will not receive this ICMP port unreachable, it will usually not be reported as dead locally either, and a host unreachable may or may not be reported by an intermediary router (some routers have a 4-hour ARP cache). Just try yourself your script on an IP address where there is no server, and see "0" being returned!

In addition, these servers' ICMP rate limiting could prevent a portion of the checks from receiving that packet depending on the number of proxies testing it and the configured rate limits (have a look at `/proc/sys/net/ipv4/icmp_rate*`).

In summary, just keep this in mind: **generic UDP checks are okay for manual debugging but are not suitable at all for anything related to production stability**. The only way to do proper UDP check is by sending a request to a service and expecting a valid response, but UDP is used with a number of unidirectional services (like syslog here).

**Comment by Clanck-cyber** [comment]
Hi, 
I try to send firewall logs to syslog using UDP. Between the syslog and the firewall I have an HAProxy cluster.
I saw that the flow arrives to the HAProxy  but when I activate debug mode there is no communication between the HAProxy  and the syslog.
I use port 2000 with UDP protocol.

My HAproxy version is : 

```
HAProxy version 2.5.5-1~bpo11+1 2022/03/14 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.5.html
Running on: Linux 5.10.0-12-amd64 #1 SMP Debian 5.10.103-1 (2022-03-07) x86_64
```

The configuration of HA proxy : 

```
global
    
      log stderr format iso local7
       # log /dev/log    local0
       # log /dev/log    local1 notice
       # chroot /var/lib/haproxy
       # stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
       # stats timeout 30s
       # user root
       # group root
       # daemon

        # Default SSL material locations
        #ca-base /etc/ssl/certs
        #crt-base /etc/ssl/private

        # See: https://ssl-config.mozilla.org/#server=haproxy&server-version=2.0.3&config=intermediate
        #ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY13>
        #ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        #ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets

defaults
        log     global
        mode    http
        option  httplog
        option  dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

log-forward sylog-loadb
# VIP of HA proxy
     dgram-bind 10.10.10.1:2000 
     bind 10.10.10.1:2000     
       # IP of my syslog server    
     log 10.10.10.2:2000  sample 1:2 local0
     log  10.10.10.3:2000 sample 2:2 local0
```

Please have you any idea about where the problem could be ? 
Thank you.
Best regards.
